src/main/kotlin/com/example/passmanager/configuration/DtoVerifierBeanPostProcessor.kt

```diff
@@ -0,0 +1,98 @@
+package com.example.passmanager.configuration
+
+import com.example.passmanager.exception.InvalidIdTypeException
+import org.bson.types.ObjectId
+import org.springframework.beans.factory.config.BeanPostProcessor
+import org.springframework.cglib.proxy.Enhancer
+import org.springframework.cglib.proxy.MethodInterceptor
+import org.springframework.stereotype.Component
+import org.springframework.web.bind.annotation.RestController
+import java.lang.reflect.Method
+import kotlin.reflect.full.declaredMemberProperties
+import kotlin.reflect.full.hasAnnotation
+import kotlin.reflect.full.primaryConstructor
+import kotlin.reflect.jvm.isAccessible
+import kotlin.reflect.jvm.jvmErasure
+
+@Component
+class DtoVerifierBeanPostProcessor : BeanPostProcessor {
+    private val beanNameToClassMap: MutableMap<String, DtoAnnotatedMethods> = mutableMapOf()
+
+    override fun postProcessBeforeInitialization(bean: Any, beanName: String): Any {
+        val beanClass = bean::class
+        if (beanClass.hasAnnotation<RestController>()) {
+            val membersWithDtoAsParam = getMethodsWithAnnotatedParams<ValidObjectIdFormat>(bean)
+            if (membersWithDtoAsParam.isNotEmpty()) {
+                beanNameToClassMap[beanName] = DtoAnnotatedMethods(beanClass.java, membersWithDtoAsParam)
+            }
+        }
+        return bean
+    }
+
+    private inline fun <reified T : Annotation> getMethodsWithAnnotatedParams(bean: Any): Set<Method> {
+        val beanClass = bean::class.java
+        return beanClass.methods.filter {
+            it.parameterTypes.any { paramType -> paramType.isAnnotationPresent(T::class.java) }
+        }.toSet()
+    }
+
+    override fun postProcessAfterInitialization(bean: Any, beanName: String): Any {
+        return beanNameToClassMap[beanName]?.let { targetMethods ->
+            createEnhancer(bean) { obj, method, args, proxy ->
+                if (method in targetMethods.methodsWithDto) {
+                    val idViolationsList = getIdViolationsList(args)
+                    if (idViolationsList.isNotEmpty()) throw InvalidIdTypeException(idViolationsList)
+                }
+                proxy.invokeSuper(obj, args ?: emptyArray())
+            }
+        } ?: bean
+    }
+
+    private fun getIdViolationsList(args: Array<Any>): List<String> {
+        val dtoParams = args.filter { it::class.hasAnnotation<ValidObjectIdFormat>() }
+        return dtoParams.flatMap { dto ->
+            getIdFieldsValues(dto)
+                .filterIsInstance<String>()
+                .filter { !ObjectId.isValid(it) }
+        }
+    }
+
+    private fun getIdFieldsValues(dto: Any): List<Any?> {
+        val dtoProperties = dto::class.declaredMemberProperties
+        return dtoProperties.filter { it.name.contains(ID, ignoreCase = true) }
```

Comment:
probably it is better to annotate properties that should be validated. because your dto may contain a parameter that is ObjectId and another parameter named ...Id but this id is from 3rd party system and may be in different format